tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -79,34 +78,44 @@ private static final class State {
         private final int size;
         private final long startTick;
         // Count is updated after samples are successfully added to the map.
-        private final ConcurrentSkipListMap<Double, WeightedSample> values;
+        private final ConcurrentSkipListMap<Double, WeightedSampleWithExemplar> values;
 
         private volatile int count;
 
+        private final ExemplarMetadataProvider<?> exemplarMetadataProvider;
+
         State(
                 double alphaNanos,
                 int size,
                 long startTick,
                 int count,
-                ConcurrentSkipListMap<Double, WeightedSample> values) {
+                ConcurrentSkipListMap<Double, WeightedSampleWithExemplar> values,
+                ExemplarMetadataProvider<?> exemplarMetadataProvider) {
             this.alphaNanos = alphaNanos;
             this.size = size;
             this.startTick = startTick;
             this.values = values;
             this.count = count;
+            this.exemplarMetadataProvider = exemplarMetadataProvider;
         }
 
         private void update(long value, long timestampNanos) {
             double itemWeight = weight(timestampNanos - startTick);
             double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
             boolean mapIsFull = count >= size;
             if (!mapIsFull || values.firstKey() < priority) {
-                addSample(priority, value, itemWeight, mapIsFull);
+                addSample(priority, value, itemWeight, mapIsFull, exemplarMetadataProvider.collect(value));
             }
         }
 
-        private void addSample(double priority, long value, double itemWeight, boolean bypassIncrement) {
-            if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight)) == null
+        private void addSample(
+                double priority,
+                long value,
+                double itemWeight,
+                boolean bypassIncrement,
+                @Nullable Object exemplarMetadata) {
+            if (values.putIfAbsent(priority, new WeightedSampleWithExemplar(value, itemWeight, exemplarMetadata))
+                            == null
```

Comment:
I think we discussed memory utilization of holding samples at one point, thinking through this out loud here:

If we assume a no-op exemplar sampler (always returns null), we've replaced the old weighted sample with a new type that's nearly identical, but has an additional Object reference.

Assuming a standard configuration on a reasonably modern JVM (e.g. G1GC + heap smaller than 32g) we should go from

before (32 bytes):
```
com.codahale.metrics.WeightedSnapshot.WeightedSample object internals:
OFF  SZ     TYPE DESCRIPTION               VALUE
  0   8          (object header: mark)     N/A
  8   4          (object header: class)    N/A
 12   4          (alignment/padding gap)   
 16   8     long WeightedSample.value      N/A
 24   8   double WeightedSample.weight     N/A
Instance size: 32 bytes
Space losses: 4 bytes internal + 0 bytes external = 4 bytes total
```

<details>
  <summary>Without compressed oops (also 32 bytes)</summary>

```
com.codahale.metrics.WeightedSnapshot.WeightedSample object internals:
OFF  SZ     TYPE DESCRIPTION               VALUE
  0   8          (object header: mark)     N/A
  8   8          (object header: class)    N/A
 16   8     long WeightedSample.value      N/A
 24   8   double WeightedSample.weight     N/A
Instance size: 32 bytes
Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```

</details>

after:
```
com.palantir.tritium.metrics.registry.WeightedSnapshotWithExemplars.WeightedSampleWithExemplar object internals:
OFF  SZ     TYPE DESCRIPTION                           VALUE
  0   8          (object header: mark)                 N/A
  8   4          (object header: class)                N/A
 12   4   Object WeightedSampleWithExemplar.metadata   N/A
 16   8     long WeightedSampleWithExemplar.value      N/A
 24   8   double WeightedSampleWithExemplar.weight     N/A
Instance size: 32 bytes
Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```

However, if we don't have compressed oops (e.g. Z1GC), the size will be impacted (40 bytes):
```
com.palantir.tritium.metrics.registry.WeightedSnapshotWithExemplars.WeightedSampleWithExemplar object internals:
OFF  SZ     TYPE DESCRIPTION                           VALUE
  0   8          (object header: mark)                 N/A
  8   8          (object header: class)                N/A
 16   8     long WeightedSampleWithExemplar.value      N/A
 24   8   double WeightedSampleWithExemplar.weight     N/A
 32   8   Object WeightedSampleWithExemplar.metadata   N/A
Instance size: 40 bytes
Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```

32->40 bytes isn't terrible in the zgc case, especially considering the map key is a boxed double, which takes an additional ~24 bytes, plus the Map.Entry object, so the 8 byte change shouldn't be terribly dramatic.

We could extract an interface and provide two implementations, one which has the metadata slot, and another which does not, eliminating that memory change for zgc services -- unclear if that's worthwhile. What do you think? @schlosna may have opinions :-)

Reply:
Thanks for the detailed analysis! My gut reaction here is that when we wire this up to witchcraft, we can just use the upstream Reservoir if we're feeding it a a No-op ExemplarMetadataProvider - so for an actual no-op case we'd have no differences in memory footprint.
For the cases with an actual provider, you guys probably know better than me how prevalent the older JVMs are, but I'm tempted to ignore the 32->40 bytes per-sample increase on the WeightedSampleWithExemplar class and accept the alignment "luck" of not having any memory footprint difference with compressed oops.

What I am more concerned about is the actual footprint of the generic type object itself (the `T exemplarMetadata`) and the timestamp, which (assuming we can store a TraceId with 64 bits and avoid a string) would be 8 bytes for the traceId + 8 bytes for the timestamp + object headers (8+4?), so potentially closer to 32 bytes with alignment. But I think we can discuss that when we have the object definition in the PR.

As far as the no-ops go (for folks not using Exemplars at all) I think using the upstream Reservoir makes more sense and then there's no issue, but curious for @schlosna's opinion!

Reply:
I assume if we want to use exemplars we'd instantiate a `com.palantir.tritium.metrics.registry.LockFreeExponentiallyDecayingReservoir` and if we didn't want exemplars we'd create a `com.codahale.metrics.LockFreeExponentiallyDecayingReservoir`. I think that is mostly reasonable, but we may need to be aware of possible megamorphic `Reservoir` callsites as a service may use multiple different `Reservoir` implementations beyond these two `LockFreeExponentiallyDecayingReservoir`, see https://shipilev.net/jvm/anatomy-quarks/16-megamorphic-virtual-calls/

Reply:
I think there's some unknown space as we begin to roll out exemplars -- we may choose to sample exemplars only when traces are sampled, so even when enabled, the memory footprint may look very similar to the state of the world today